# F21DL: Coursework Part 4 and 5

## Preparation

In [1]:
import tensorflow as tf
from tensorflow import keras
import numpy as np
import sklearn

%matplotlib inline
import matplotlib as mpl
import matplotlib.pyplot as plt
mpl.rc('axes', labelsize=14)
mpl.rc('xtick', labelsize=12)
mpl.rc('ytick', labelsize=12)

In [2]:
# Training data
with np.load("/content/drive/MyDrive/Colab Notebooks/F21DL/kmnist/k49-train-imgs.npz") as xTrain:
    X_train = xTrain['arr_0']

# Training labels
with np.load("/content/drive/MyDrive/Colab Notebooks/F21DL/kmnist/k49-train-labels.npz") as yTrain:
    y_train = yTrain["arr_0"]

# Testing data 
with np.load("/content/drive/MyDrive/Colab Notebooks/F21DL/kmnist/k49-test-imgs.npz") as xTest:
    X_test = xTest["arr_0"]

#Testing labels
with np.load("/content/drive/MyDrive/Colab Notebooks/F21DL/kmnist/k49-test-labels.npz") as yTest:
    y_test = yTest["arr_0"]

## Multilayer Perceptron

In [3]:
X_valid, X_train = X_train[:5000] / 255., X_train[5000:] / 255.
y_valid, y_train = y_train[:5000], y_train[5000:]
X_test = X_test / 255.

In [4]:
keras.backend.clear_session()
np.random.seed(42)
tf.random.set_seed(42)

In [5]:
model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[28, 28])) # first layer convert each input image into a 1D array
model.add(keras.layers.Dense(300, activation="relu")) #hidden layer with 300 neurons useing the ReLU activation function.
model.add(keras.layers.Dense(100, activation="relu")) #hidden layer with 100 neurons using the ReLU
model.add(keras.layers.Dense(49, activation="softmax")) #output layer with 49 neurons (one per class), using the softmax 

In [6]:
model.compile(loss=keras.losses.SparseCategoricalCrossentropy(),
              optimizer=keras.optimizers.SGD(), #"sgd" simply means that we will train the model using simple Stochastic Gradient Descent
              metrics=[keras.metrics.SparseCategoricalAccuracy()])

In [7]:
history = model.fit(X_train, y_train, epochs=10, validation_data=(X_valid, y_valid))

Epoch 1/10
7106/7106 [==============================] - 17s 2ms/step - loss: 1.6159 - sparse_categorical_accuracy: 0.6088 - val_loss: 1.0935 - val_sparse_categorical_accuracy: 0.7164
Epoch 2/10
7106/7106 [==============================] - 14s 2ms/step - loss: 0.9485 - sparse_categorical_accuracy: 0.7608 - val_loss: 0.8503 - val_sparse_categorical_accuracy: 0.7736
Epoch 3/10
7106/7106 [==============================] - 14s 2ms/step - loss: 0.7417 - sparse_categorical_accuracy: 0.8103 - val_loss: 0.6851 - val_sparse_categorical_accuracy: 0.8190
Epoch 4/10
7106/7106 [==============================] - 14s 2ms/step - loss: 0.6214 - sparse_categorical_accuracy: 0.8394 - val_loss: 0.6508 - val_sparse_categorical_accuracy: 0.8256
Epoch 5/10
7106/7106 [==============================] - 14s 2ms/step - loss: 0.5418 - sparse_categorical_accuracy: 0.8588 - val_loss: 0.5541 - val_sparse_categorical_accuracy: 0.8530
Epoch 6/10
7106/7106 [==============================] - 14s 2ms/step - loss: 0.4842 -

In [8]:
results_base_train = model.evaluate(X_train, y_train)

7106/7106 [==============================] - 13s 2ms/step - loss: 0.3334 - sparse_categorical_accuracy: 0.9113


In [9]:
results_base_train

[0.3334272503852844, 0.9113144278526306]

In [10]:
results_base_test = model.evaluate(X_test, y_test)

1205/1205 [==============================] - 2s 2ms/step - loss: 0.7779 - sparse_categorical_accuracy: 0.8037


**The following accuracies will be used as the base case for all comparisons in the modification section.**

In [11]:
print(f"Training Accuracy: {results_base_train[1]:.6f}")
print(f"Testing  Accuracy: {results_base_test[1]:.6f}")

Training Accuracy: 0.911314
Testing  Accuracy: 0.803694


## Parameter Modifications

### Change Activation Functions

In [12]:
keras.backend.clear_session()

In [13]:
model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[28, 28])) 
model.add(keras.layers.Dense(300, activation="sigmoid"))
model.add(keras.layers.Dense(100, activation="tanh"))
model.add(keras.layers.Dense(49, activation="relu"))

In [14]:
model.compile(loss="sparse_categorical_crossentropy",
              optimizer="sgd", #"sgd" simply means that we will train the model using simple Stochastic Gradient Descent
              metrics=["accuracy"])

In [15]:
history = model.fit(X_train, y_train, epochs=10, validation_data=(X_valid, y_valid))

Epoch 1/10
7106/7106 [==============================] - 16s 2ms/step - loss: 7.9369 - accuracy: 0.0227 - val_loss: 7.6677 - val_accuracy: 0.0216
Epoch 2/10
7106/7106 [==============================] - 15s 2ms/step - loss: 8.3635 - accuracy: 0.0259 - val_loss: 8.0537 - val_accuracy: 0.0216
Epoch 3/10
7106/7106 [==============================] - 15s 2ms/step - loss: 7.8706 - accuracy: 0.0259 - val_loss: 7.7169 - val_accuracy: 0.0216
Epoch 4/10
7106/7106 [==============================] - 15s 2ms/step - loss: 7.7050 - accuracy: 0.0259 - val_loss: 7.7194 - val_accuracy: 0.0216
Epoch 5/10
7106/7106 [==============================] - 15s 2ms/step - loss: 7.7154 - accuracy: 0.0259 - val_loss: 7.7194 - val_accuracy: 0.0216
Epoch 6/10
7106/7106 [==============================] - 15s 2ms/step - loss: 7.7154 - accuracy: 0.0259 - val_loss: 7.7194 - val_accuracy: 0.0216
Epoch 7/10
7106/7106 [==============================] - 16s 2ms/step - loss: 7.7154 - accuracy: 0.0259 - val_loss: 7.7194 - val_ac

In [16]:
results_act_train = model.evaluate(X_train, y_train)
results_act_test = model.evaluate(X_test, y_test)
print(f"Training Accuracy: {results_act_train[1]:.6f}")
print(f"Testing  Accuracy: {results_act_test[1]:.6f}")

1205/1205 [==============================] - 2s 2ms/step - loss: 7.7164 - accuracy: 0.0259
Training Accuracy: 0.025914
Testing  Accuracy: 0.025942


#### Observations
1. Training accuracy **decreased by 0.8854.**
2. Testing accuracy   **decreased by 0.7778.**

### Change Number and Size of Layers

In [17]:
keras.backend.clear_session()

model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[28, 28])) 
model.add(keras.layers.Dense(600, activation="relu"))
model.add(keras.layers.Dense(300, activation="relu"))
model.add(keras.layers.Dense(200, activation="relu"))
model.add(keras.layers.Dense(100, activation="relu"))
model.add(keras.layers.Dense(49, activation="softmax"))

model.compile(loss="sparse_categorical_crossentropy",
              optimizer="sgd", #"sgd" simply means that we will train the model using simple Stochastic Gradient Descent
              metrics=["accuracy"])

history = model.fit(X_train, y_train, epochs=10, validation_data=(X_valid, y_valid))

results_lay_train = model.evaluate(X_train, y_train)
results_lay_test = model.evaluate(X_test, y_test)

Epoch 1/10
7106/7106 [==============================] - 16s 2ms/step - loss: 1.5205 - accuracy: 0.6224 - val_loss: 0.9268 - val_accuracy: 0.7500
Epoch 2/10
7106/7106 [==============================] - 15s 2ms/step - loss: 0.7211 - accuracy: 0.8102 - val_loss: 0.6827 - val_accuracy: 0.8150
Epoch 3/10
7106/7106 [==============================] - 15s 2ms/step - loss: 0.5259 - accuracy: 0.8585 - val_loss: 0.4722 - val_accuracy: 0.8728
Epoch 4/10
7106/7106 [==============================] - 15s 2ms/step - loss: 0.4229 - accuracy: 0.8856 - val_loss: 0.6864 - val_accuracy: 0.8024
Epoch 5/10
7106/7106 [==============================] - 15s 2ms/step - loss: 0.3564 - accuracy: 0.9025 - val_loss: 0.4609 - val_accuracy: 0.8708
Epoch 6/10
7106/7106 [==============================] - 15s 2ms/step - loss: 0.3066 - accuracy: 0.9158 - val_loss: 0.3755 - val_accuracy: 0.8970
Epoch 7/10
7106/7106 [==============================] - 15s 2ms/step - loss: 0.2680 - accuracy: 0.9261 - val_loss: 0.3668 - val_ac

In [18]:
print(f"Training Accuracy: {results_lay_train[1]:.6f}")
print(f"Testing  Accuracy: {results_lay_test[1]:.6f}")

Training Accuracy: 0.951048
Testing  Accuracy: 0.837860


#### Observations
1. Training accuracy **increased by 0.039734.**
2. Testing accuracy **increased by 0.034166.**

### Change Learning Rate

In [21]:
keras.backend.clear_session()

model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[28, 28])) 
model.add(keras.layers.Dense(300, activation="relu"))
model.add(keras.layers.Dense(100, activation="relu"))
model.add(keras.layers.Dense(49, activation="softmax"))

model.compile(loss="sparse_categorical_crossentropy",
              optimizer=keras.optimizers.SGD(learning_rate=2), #changing lr from default 0.01
              metrics=["accuracy"])

history = model.fit(X_train, y_train, epochs=10, validation_data=(X_valid, y_valid))

results_lr_train = model.evaluate(X_train, y_train)
results_lr_test = model.evaluate(X_test, y_test)

Epoch 1/10
7106/7106 [==============================] - 15s 2ms/step - loss: 3.8260 - accuracy: 0.0261 - val_loss: 3.8105 - val_accuracy: 0.0246
Epoch 2/10
7106/7106 [==============================] - 14s 2ms/step - loss: 3.7649 - accuracy: 0.0362 - val_loss: 3.8182 - val_accuracy: 0.0216
Epoch 3/10
7106/7106 [==============================] - 14s 2ms/step - loss: 3.8107 - accuracy: 0.0259 - val_loss: 3.8143 - val_accuracy: 0.0258
Epoch 4/10
7106/7106 [==============================] - 14s 2ms/step - loss: 3.8107 - accuracy: 0.0253 - val_loss: 3.8225 - val_accuracy: 0.0232
Epoch 5/10
7106/7106 [==============================] - 15s 2ms/step - loss: 3.8107 - accuracy: 0.0260 - val_loss: 3.8090 - val_accuracy: 0.0228
Epoch 6/10
7106/7106 [==============================] - 14s 2ms/step - loss: 3.8113 - accuracy: 0.0262 - val_loss: 3.8145 - val_accuracy: 0.0264
Epoch 7/10
7106/7106 [==============================] - 14s 2ms/step - loss: 3.8108 - accuracy: 0.0263 - val_loss: 3.8150 - val_ac

In [23]:
print(f"Training Accuracy: {results_lr_train[1]:.6f}")
print(f"Testing  Accuracy: {results_lr_test[1]:.6f}")

Training Accuracy: 0.025809
Testing  Accuracy: 0.025942


#### Observations
1. Training accuracy **decreased by 0.885505.**
2. Testing accuracy **decreased by 0.777778.**
3. This is surprisingly the same resut as when i changed the activation functions. 

### Change Momentum

In [24]:
keras.backend.clear_session()

model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[28, 28])) 
model.add(keras.layers.Dense(300, activation="relu"))
model.add(keras.layers.Dense(100, activation="relu"))
model.add(keras.layers.Dense(49, activation="softmax"))

model.compile(loss="sparse_categorical_crossentropy",
              optimizer=keras.optimizers.SGD(momentum=0.9), #changing mom from default 0
              metrics=["accuracy"])

history = model.fit(X_train, y_train, epochs=10, validation_data=(X_valid, y_valid))

results_mom_train = model.evaluate(X_train, y_train)
results_mom_test = model.evaluate(X_test, y_test)

Epoch 1/10
7106/7106 [==============================] - 15s 2ms/step - loss: 0.8423 - accuracy: 0.7781 - val_loss: 0.5081 - val_accuracy: 0.8658
Epoch 2/10
7106/7106 [==============================] - 15s 2ms/step - loss: 0.4403 - accuracy: 0.8779 - val_loss: 0.4222 - val_accuracy: 0.8828
Epoch 3/10
7106/7106 [==============================] - 15s 2ms/step - loss: 0.3471 - accuracy: 0.9023 - val_loss: 0.3796 - val_accuracy: 0.8986
Epoch 4/10
7106/7106 [==============================] - 15s 2ms/step - loss: 0.2929 - accuracy: 0.9167 - val_loss: 0.3675 - val_accuracy: 0.9012
Epoch 5/10
7106/7106 [==============================] - 15s 2ms/step - loss: 0.2579 - accuracy: 0.9255 - val_loss: 0.3592 - val_accuracy: 0.8992
Epoch 6/10
7106/7106 [==============================] - 15s 2ms/step - loss: 0.2288 - accuracy: 0.9327 - val_loss: 0.3643 - val_accuracy: 0.9072
Epoch 7/10
7106/7106 [==============================] - 15s 2ms/step - loss: 0.2063 - accuracy: 0.9390 - val_loss: 0.3672 - val_ac

In [25]:
print(f"Training Accuracy: {results_mom_train[1]:.6f}")
print(f"Testing  Accuracy: {results_mom_test[1]:.6f}")

Training Accuracy: 0.959985
Testing  Accuracy: 0.840558


#### Observations
1. Training accuracy **increased by 0.048671.**
2. Testing accuracy **increased by 0.036864.**

### Change Epochs

In [26]:
keras.backend.clear_session()

model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[28, 28])) 
model.add(keras.layers.Dense(300, activation="relu"))
model.add(keras.layers.Dense(100, activation="relu"))
model.add(keras.layers.Dense(49, activation="softmax"))

model.compile(loss="sparse_categorical_crossentropy",
              optimizer=keras.optimizers.SGD(),
              metrics=["accuracy"])

history = model.fit(X_train, y_train, epochs=30, validation_data=(X_valid, y_valid))

results_epo_train = model.evaluate(X_train, y_train)
results_epo_test = model.evaluate(X_test, y_test)

Epoch 1/30
7106/7106 [==============================] - 14s 2ms/step - loss: 1.6322 - accuracy: 0.6021 - val_loss: 1.0943 - val_accuracy: 0.7218
Epoch 2/30
7106/7106 [==============================] - 14s 2ms/step - loss: 0.9481 - accuracy: 0.7600 - val_loss: 0.8489 - val_accuracy: 0.7708
Epoch 3/30
7106/7106 [==============================] - 14s 2ms/step - loss: 0.7384 - accuracy: 0.8100 - val_loss: 0.6784 - val_accuracy: 0.8244
Epoch 4/30
7106/7106 [==============================] - 14s 2ms/step - loss: 0.6175 - accuracy: 0.8395 - val_loss: 0.6552 - val_accuracy: 0.8194
Epoch 5/30
7106/7106 [==============================] - 14s 2ms/step - loss: 0.5384 - accuracy: 0.8594 - val_loss: 0.5402 - val_accuracy: 0.8520
Epoch 6/30
7106/7106 [==============================] - 14s 2ms/step - loss: 0.4812 - accuracy: 0.8735 - val_loss: 0.5009 - val_accuracy: 0.8644
Epoch 7/30
7106/7106 [==============================] - 14s 2ms/step - loss: 0.4383 - accuracy: 0.8839 - val_loss: 0.4749 - val_ac

In [27]:
print(f"Training Accuracy: {results_epo_train[1]:.6f}")
print(f"Testing  Accuracy: {results_epo_test[1]:.6f}")

Training Accuracy: 0.964238
Testing  Accuracy: 0.833735


#### Observations
1. Training accuracy **increased by 0.052924.**
2. Testing accuracy **increased by 0.030041.**

### Change Validation Threshold

In [40]:
from keras.layers import ReLU

In [31]:
keras.backend.clear_session()

model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[28, 28])) 
# changing relu threshold values from default 0
model.add(keras.layers.Dense(300, activation=ReLU(threshold=0.5)))
model.add(keras.layers.Dense(100, activation=ReLU(threshold=4)))
model.add(keras.layers.Dense(49, activation="softmax"))

model.compile(loss="sparse_categorical_crossentropy",
              optimizer=keras.optimizers.SGD(),
              metrics=["accuracy"])

history = model.fit(X_train, y_train, epochs=10, validation_data=(X_valid, y_valid))

results_thr_train = model.evaluate(X_train, y_train)
results_thr_test = model.evaluate(X_test, y_test)

Epoch 1/10
7106/7106 [==============================] - 15s 2ms/step - loss: 3.8381 - accuracy: 0.0254 - val_loss: 3.8110 - val_accuracy: 0.0264
Epoch 2/10
7106/7106 [==============================] - 15s 2ms/step - loss: 3.8057 - accuracy: 0.0251 - val_loss: 3.7994 - val_accuracy: 0.0232
Epoch 3/10
7106/7106 [==============================] - 15s 2ms/step - loss: 3.7997 - accuracy: 0.0250 - val_loss: 3.7959 - val_accuracy: 0.0232
Epoch 4/10
7106/7106 [==============================] - 15s 2ms/step - loss: 3.7977 - accuracy: 0.0250 - val_loss: 3.7946 - val_accuracy: 0.0216
Epoch 5/10
7106/7106 [==============================] - 15s 2ms/step - loss: 3.7968 - accuracy: 0.0253 - val_loss: 3.7940 - val_accuracy: 0.0246
Epoch 6/10
7106/7106 [==============================] - 15s 2ms/step - loss: 3.7963 - accuracy: 0.0251 - val_loss: 3.7936 - val_accuracy: 0.0264
Epoch 7/10
7106/7106 [==============================] - 15s 2ms/step - loss: 3.7960 - accuracy: 0.0251 - val_loss: 3.7935 - val_ac

In [32]:
print(f"Training Accuracy: {results_thr_train[1]:.6f}")
print(f"Testing  Accuracy: {results_thr_test[1]:.6f}")

Training Accuracy: 0.025809
Testing  Accuracy: 0.025942


#### Observations
1. Training accuracy **decreased by 0.885505.**
2. Testing accuracy **decreased by 0.777752.**
3. Once again, these new (bad) values are same as when I changed the activation functions or Learning Rate.

### Conclusions 
1. Given the high accuracies on some of the models, and the fact that Hyperparameter tuning will increase even it even more, it appears that this dataset **is linearly separable** when using neural networks.
2. As for neural network's capacity to generalize to new data, in my experiments they have been just as good as with training data or only few percentage points (~10) worse. I am inclined to believe that with hyperparameter tuning the **neural networks will generalize to new data even better.**
3. Lastly, as seen in the observations for 'Change Activation' section, functions can make a **huge difference in accuracies, i.e., network's ability to learn and predict classes.** For instance, it appears important that we use **softmax activation** in the last layer. Apart from that, ReLU activations seemed to perform better than the sigmoid and tanh functions, although this observation is just from a single run and not a detailed study.

## Convolutional Neural Network

In [5]:
print(X_train.shape)
print(X_test.shape)

(232365, 28, 28)
(38547, 28, 28)


In [7]:
x_train = X_train.reshape(232365, 28, 28, 1)
x_test = X_test.reshape(38547, 28, 28, 1)
x_train.shape #28*28=784

(232365, 28, 28, 1)

In [20]:
keras.backend.clear_session()
model = tf.keras.models.Sequential([
  tf.keras.layers.Conv2D(64, (3,3), activation='relu', input_shape=(28, 28, 1)),
  tf.keras.layers.MaxPooling2D(2, 2),
  tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
  tf.keras.layers.MaxPooling2D(2,2),
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(49, activation='softmax')
])
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
history = model.fit(x_train, y_train, epochs=5)
results_train = model.evaluate(x_train, y_train)
results_test =  model.evaluate(x_test, y_test)

Epoch 1/5
7262/7262 [==============================] - 29s 4ms/step - loss: 0.5571 - accuracy: 0.8592
Epoch 2/5
7262/7262 [==============================] - 28s 4ms/step - loss: 0.2502 - accuracy: 0.9300
Epoch 3/5
7262/7262 [==============================] - 28s 4ms/step - loss: 0.2005 - accuracy: 0.9435
Epoch 4/5
7262/7262 [==============================] - 28s 4ms/step - loss: 0.1708 - accuracy: 0.9512
Epoch 5/5
1205/1205 [==============================] - 3s 3ms/step - loss: 0.4854 - accuracy: 0.8921


**Note: The following accuracies will be used as the baseline to compare all the experiments with.**

In [21]:
print(f"Training Accuracy: {results_train[1]:.4f}")
print(f"Testing  Accuracy: {results_test[1]:.4f}")

Training Accuracy: 0.9628
Testing  Accuracy: 0.8921


### Change Activation Functions

In [22]:
keras.backend.clear_session()
model = tf.keras.models.Sequential([
  tf.keras.layers.Conv2D(64, (3,3), activation='sigmoid', input_shape=(28, 28, 1)),
  tf.keras.layers.MaxPooling2D(2, 2),
  tf.keras.layers.Conv2D(64, (3,3), activation='tanh'),
  tf.keras.layers.MaxPooling2D(2,2),
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(49, activation='softmax')
])
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
history = model.fit(x_train, y_train, epochs=5)
cnn_results_act_train = model.evaluate(x_train, y_train)
cnn_results_act_test =  model.evaluate(x_test, y_test)

Epoch 1/5
7262/7262 [==============================] - 29s 4ms/step - loss: 0.5039 - accuracy: 0.8651
Epoch 2/5
7262/7262 [==============================] - 28s 4ms/step - loss: 0.2637 - accuracy: 0.9280
Epoch 3/5
7262/7262 [==============================] - 28s 4ms/step - loss: 0.2160 - accuracy: 0.9403
Epoch 4/5
7262/7262 [==============================] - 28s 4ms/step - loss: 0.1896 - accuracy: 0.9478
Epoch 5/5
1205/1205 [==============================] - 3s 3ms/step - loss: 0.5007 - accuracy: 0.8904


In [23]:
print(f"Training Accuracy: {cnn_results_act_train[1]:.4f}")
print(f"Testing  Accuracy: {cnn_results_act_test[1]:.4f}")

Training Accuracy: 0.9601
Testing  Accuracy: 0.8904


#### Observations
1. Training accuracy **decreased by 0.0027.**
2. Testing accuracy **decreased by 0.0017.**

### Change Number and Size of Layers

In [31]:
# Added more Cov2D layers, changed filter size, added another connected layer

In [32]:
keras.backend.clear_session()
model = tf.keras.models.Sequential([
  tf.keras.layers.Conv2D(64, (3,3), activation='relu', input_shape=(28, 28, 1)),
  tf.keras.layers.MaxPooling2D(2, 2),
  tf.keras.layers.Conv2D(64, (2,2), activation='relu'),
  tf.keras.layers.MaxPooling2D(2, 2),
  tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
  tf.keras.layers.MaxPooling2D(2,2),
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(200, activation='relu'),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(49, activation='softmax')
])
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
history = model.fit(x_train, y_train, epochs=5)
cnn_results_lay_train = model.evaluate(x_train, y_train)
cnn_results_lay_test =  model.evaluate(x_test, y_test)

Epoch 1/5
7262/7262 [==============================] - 34s 5ms/step - loss: 0.5303 - accuracy: 0.8580
Epoch 2/5
7262/7262 [==============================] - 33s 5ms/step - loss: 0.2465 - accuracy: 0.9312
Epoch 3/5
7262/7262 [==============================] - 33s 5ms/step - loss: 0.2057 - accuracy: 0.9425
Epoch 4/5
7262/7262 [==============================] - 33s 5ms/step - loss: 0.1869 - accuracy: 0.9474
Epoch 5/5
1205/1205 [==============================] - 4s 3ms/step - loss: 0.4001 - accuracy: 0.8989


In [33]:
print(f"Training Accuracy: {cnn_results_lay_train[1]:.4f}")
print(f"Testing  Accuracy: {cnn_results_lay_test[1]:.4f}")

Training Accuracy: 0.9563
Testing  Accuracy: 0.8989


#### Observations
1. Training accuracy **decreased by 0.0065.**
2. Testing accuracy **increased by 0.0068.**
3. Although accuracy on training set lowered, it is marginal and changes to model actually increased accuracy for new data.

### Change Learning Rate

In [41]:
keras.backend.clear_session()
model = tf.keras.models.Sequential([
  tf.keras.layers.Conv2D(64, (3,3), activation='relu', input_shape=(28, 28, 1)),
  tf.keras.layers.MaxPooling2D(2, 2),
  tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
  tf.keras.layers.MaxPooling2D(2,2),
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(49, activation='softmax')
])
model.compile(optimizer=keras.optimizers.Adam(learning_rate=10), loss='sparse_categorical_crossentropy', metrics=['accuracy'])
history = model.fit(x_train, y_train, epochs=5)
cnn_results_lr_train = model.evaluate(x_train, y_train)
cnn_results_lr_test =  model.evaluate(x_test, y_test)

Epoch 1/5
7262/7262 [==============================] - 29s 4ms/step - loss: 1301879296.0000 - accuracy: 0.0238
Epoch 2/5
7262/7262 [==============================] - 29s 4ms/step - loss: 7.3680 - accuracy: 0.0244
Epoch 3/5
7262/7262 [==============================] - 29s 4ms/step - loss: 7.4740 - accuracy: 0.0237
Epoch 4/5
7262/7262 [==============================] - 29s 4ms/step - loss: 7.4111 - accuracy: 0.0242
Epoch 5/5
1205/1205 [==============================] - 3s 3ms/step - loss: 9.2904 - accuracy: 0.0259


In [42]:
print(f"Training Accuracy: {cnn_results_lr_train[1]:.4f}")
print(f"Testing  Accuracy: {cnn_results_lr_test[1]:.4f}")

Training Accuracy: 0.0258
Testing  Accuracy: 0.0259


#### Observations
1. The loss at end of Epoch 1 is **MASSIVE at 1301879296.**
2. Training accuracy **decreased by 0.9370.**
3. Testing accuracy **decreased by 0.8662.**
4. This is expected behaviour. A much higher learning rate means that we can "skip" over the local minima because of "longer jumps".

### Change Momentum

In [48]:
# Momentum appears to be the beta1 value in the Adam Optimizer.
# Lowered it from default 0.9 to 0.1
# https://stackoverflow.com/questions/47168616/is-there-a-momentum-option-for-adam-optimizer-in-keras

In [46]:
keras.backend.clear_session()
model = tf.keras.models.Sequential([
  tf.keras.layers.Conv2D(64, (3,3), activation='relu', input_shape=(28, 28, 1)),
  tf.keras.layers.MaxPooling2D(2, 2),
  tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
  tf.keras.layers.MaxPooling2D(2,2),
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(49, activation='softmax')
])
model.compile(optimizer=keras.optimizers.Adam(beta_1=0.1), loss='sparse_categorical_crossentropy', metrics=['accuracy'])
history = model.fit(x_train, y_train, epochs=5)
cnn_results_mom_train = model.evaluate(x_train, y_train)
cnn_results_mom_test =  model.evaluate(x_test, y_test)

Epoch 1/5
7262/7262 [==============================] - 29s 4ms/step - loss: 0.5220 - accuracy: 0.8660
Epoch 2/5
7262/7262 [==============================] - 28s 4ms/step - loss: 0.2242 - accuracy: 0.9378
Epoch 3/5
7262/7262 [==============================] - 28s 4ms/step - loss: 0.1760 - accuracy: 0.9504
Epoch 4/5
7262/7262 [==============================] - 28s 4ms/step - loss: 0.1492 - accuracy: 0.9574
Epoch 5/5
1205/1205 [==============================] - 3s 3ms/step - loss: 0.4261 - accuracy: 0.9054


In [47]:
print(f"Training Accuracy: {cnn_results_mom_train[1]:.4f}")
print(f"Testing  Accuracy: {cnn_results_mom_test[1]:.4f}")

Training Accuracy: 0.9721
Testing  Accuracy: 0.9054


#### Observations
1. Training accuracy **increased by 0.0093.**
2. Testing accuracy **increased by 0.0133.**

### Change Epochs

In [49]:
keras.backend.clear_session()
model = tf.keras.models.Sequential([
  tf.keras.layers.Conv2D(64, (3,3), activation='relu', input_shape=(28, 28, 1)),
  tf.keras.layers.MaxPooling2D(2, 2),
  tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
  tf.keras.layers.MaxPooling2D(2,2),
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(49, activation='softmax')
])
model.compile(optimizer=keras.optimizers.Adam(), loss='sparse_categorical_crossentropy', metrics=['accuracy'])
history = model.fit(x_train, y_train, epochs=20)
cnn_results_epo_train = model.evaluate(x_train, y_train)
cnn_results_epo_test =  model.evaluate(x_test, y_test)

Epoch 1/20
7262/7262 [==============================] - 29s 4ms/step - loss: 0.5232 - accuracy: 0.8646
Epoch 2/20
7262/7262 [==============================] - 28s 4ms/step - loss: 0.2407 - accuracy: 0.9323
Epoch 3/20
7262/7262 [==============================] - 28s 4ms/step - loss: 0.1918 - accuracy: 0.9451
Epoch 4/20
7262/7262 [==============================] - 28s 4ms/step - loss: 0.1658 - accuracy: 0.9525
Epoch 5/20
7262/7262 [==============================] - 29s 4ms/step - loss: 0.1488 - accuracy: 0.9570
Epoch 6/20
7262/7262 [==============================] - 28s 4ms/step - loss: 0.1354 - accuracy: 0.9602
Epoch 7/20
7262/7262 [==============================] - 28s 4ms/step - loss: 0.1267 - accuracy: 0.9629
Epoch 8/20
7262/7262 [==============================] - 28s 4ms/step - loss: 0.1194 - accuracy: 0.9649
Epoch 9/20
7262/7262 [==============================] - 28s 4ms/step - loss: 0.1124 - accuracy: 0.9671
Epoch 10/20
7262/7262 [==============================] - 28s 4ms/step - l

In [50]:
print(f"Training Accuracy: {cnn_results_epo_train[1]:.4f}")
print(f"Testing  Accuracy: {cnn_results_epo_test[1]:.4f}")

Training Accuracy: 0.9823
Testing  Accuracy: 0.8967


#### Observations
1. Training accuracy **increased by 0.0195.**
2. Testing accuracy **increased by 0.0046.**

### Change Validation Threshold

In [51]:
keras.backend.clear_session()
model = tf.keras.models.Sequential([
  tf.keras.layers.Conv2D(64, (3,3), activation=ReLU(threshold=5), input_shape=(28, 28, 1)),
  tf.keras.layers.MaxPooling2D(2, 2),
  tf.keras.layers.Conv2D(64, (3,3), activation=ReLU(threshold=5)),
  tf.keras.layers.MaxPooling2D(2,2),
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(128, activation=ReLU(threshold=5)),
  tf.keras.layers.Dense(49, activation='softmax')
])
model.compile(optimizer=keras.optimizers.Adam(), loss='sparse_categorical_crossentropy', metrics=['accuracy'])
history = model.fit(x_train, y_train, epochs=5)
cnn_results_val_train = model.evaluate(x_train, y_train)
cnn_results_val_test =  model.evaluate(x_test, y_test)

Epoch 1/5
7262/7262 [==============================] - 29s 4ms/step - loss: 1.2867 - accuracy: 0.6707
Epoch 2/5
7262/7262 [==============================] - 29s 4ms/step - loss: 0.6125 - accuracy: 0.8348
Epoch 3/5
7262/7262 [==============================] - 29s 4ms/step - loss: 0.5755 - accuracy: 0.8473
Epoch 4/5
7262/7262 [==============================] - 29s 4ms/step - loss: 0.6094 - accuracy: 0.8366
Epoch 5/5
1205/1205 [==============================] - 3s 3ms/step - loss: 1.4316 - accuracy: 0.6467


In [52]:
print(f"Training Accuracy: {cnn_results_val_train[1]:.4f}")
print(f"Testing  Accuracy: {cnn_results_val_test[1]:.4f}")

Training Accuracy: 0.7591
Testing  Accuracy: 0.6467


#### Observations
1. Training accuracy **decreased by 0.2037.**
2. Testing accuracy **decreased by 0.2454.**

### Conclusions 
1. Just like with MLP, given the high accuracies on some of the models, and the fact that Hyperparameter tuning will increase it even more, it appears that this dataset **is linearly separable** when using CNN.
2. As for CNN's capacity to generalize to new data, in my experiments they have been just as good as with training data or only few percentage points (~10) worse. I am inclined to believe that with hyperparameter tuning the **CNN will generalize to new data even better.**
3. Lastly, as seen in the observations for 'Change Activation' section, functions can make a ** difference in accuracies. It may not appear significant in my experiment, but that is just one random run which certainly did make the model worse so I must say activation functions are important and can have varying degrees of effect on the CNN.**

## Part 5

### **Research Question**

The CNN implementation was one of the most informative and practical parts of the portfolio. It has potential to perform much better on image data sets such as mine. Note that the dataset I have used is the **Kuzushiji-49** Dataset which has images of old style Japanese characters. This has gotten me interested in a very particular application:

**"How can CNN Hyperparameters be tuned to perform OCR on old Japanese physical literature in order to digitize them?"**

It would also be interesting to see how well such a model would perfom on the newer, standardized Japanese script used now. This could have particular significance in developing applications for NLP using text, as well as translations to for Japanese media which has obvious real-life uses for instance in the case of tourists.

### **Research Answer**

While I do not have a detailed solution to provide, I have a general idea.
<br><br>Given that our CNN can already reach accuracies of up 90% on test data, it is not unthinkable that with better training it could achieve close to 100% as well. So the solution lies in hyperparameter tuning. <br>
I would approach that problem with a Grid Search for the parameters. This would involve limiting our parameter search space to a fixed range, so that it is not infinite, and then running repeated experiments to find best mean values that give us the best parameters. <br><br>
Once we are comfortable with our model, we can start capturing live data through camera feeds which would then be used for OCR or translation purposes. 